File: python/Sorting Algorithms/mergeSort.py
def mergeSort(collection: list) -> list:
    """Examples:
    >>> mergeSort([0, 5, 3, 2, 2])
    [0, 2, 2, 3, 5]
    >>> mergeSort([])
    []
    >>> mergeSort([-2, -5, -45])
    [-45, -5, -2]
    """

    def merge(left: list, right: list) -> list:
        """merge left and right
        :param left: left collection
        :param right: right collection
        :return: merge result
        """

        def _merge():
            while left and right:
                yield (left if left[0] <= right[0] else right).pop(0)
            yield from left
            yield from right

        return list(_merge())

    if len(collection) <= 1:
        return collection
    mid = len(collection) // 2
    return merge(mergeSort(collection[:mid]), mergeSort(collection[mid:]))


if __name__ == "__main__":
    import doctest

    doctest.testmod()
    user_input = input("Enter numbers separated by a comma:\n").strip()
    unsorted = [int(item) for item in user_input.split(",")]
    print(*mergeSort(unsorted), sep=",")

File: cpp/sorting/merge_sort.cpp
#include <iostream>
using namespace std;

void merge(int *a, int *x, int *y, int s, int e)
{
    int i = s;

    int mid = (s + e) / 2;

    int j = mid + 1;	

    int k = s;

    while (i <= mid && j <= e)
    {
        if (x[i] < y[j])
        {
            a[k] = x[i];
            k++;
            i++;
        }
        else
        {
            a[k] = y[j];
            k++;
            j++;
        }
    }

    while (i <= mid)
    {
        a[k] = x[i];
        i++;
        k++;
    }

    while (j <= e)
    {
        a[k] = y[j];
        k++;
        j++;
    }

    return;

}

void mergesort(int *a, int s, int e)
{

    if (s == e)
    {
        return;
    }


    int x[100];
    int y[100];

    int mid = (s + e) / 2;

    for (int i = s; i <= mid; i++)
    {
        x[i] = a[i];
    }

    for (int i = mid + 1; i <= e; i++)
    {
        y[i] = a[i];
    }

    mergesort(x, s, mid);
    mergesort(y, mid + 1, e);
  
    merge(a, x, y, s, e);

}

int main()
{
    int n;
    cin >> n;
    int a[n];

    for(int i=0; i<n; i++){
       cin >> a[i];
    }

    mergesort(a, 0, n - 1);

    for (int i = 0; i < n; i++)
    {
        cout << a[i] << " ";
    }

    return 0;
}
